src/lib/non_empty_list/non_empty_list.ml

```diff
@@ -0,0 +1,46 @@
+open Core_kernel
+
+(* A non-empty list is a tuple of the head and the rest (as a list) *)
+type 'a t = 'a * 'a list [@@deriving sexp, compare, eq, hash]
+
+let init x xs = (x, xs)
+
+let singleton x = (x, [])
+
+let uncons = Fn.id
+
+let cons (x, xs) x' = (x', x :: xs)
```

Comment:
shouldn't the list be the second argument of cons?

Reply:
+1 what @cmr said.